pkg/etcd/stream.go

```diff
@@ -42,16 +43,27 @@ func NewStream(logger kitlog.Logger, client etcdGetter, opt StreamOptions) (<-ch
 		defer cancel()
 		defer wg.Done()
 
-		logger.Log("event", "watch.start")
-		for resp := range client.Watch(opt.Ctx, "/", clientv3.WithPrefix()) {
+	Watch:
```

Comment:
Is this genuinely the recommended way to do flow control in Go?

Reply:
Go doesn't have `do {} while` so it's not possible to express this without duplicating your initialisation without the gotos :( I wish I had a better way but every way I rewrite it gets worse.

Please do give it a go though, this is just the least bad of the approaches I've found.

Reply:
I was thinking maybe you could have this stuff below live in its own function, and signal that it's completed via a channel? But I haven't actually tried this out.

Reply:
I would try it out- I spent a while trying to rewrite the poll without goto's a few weeks back and didn't get anywhere with it. A second opinion on whether it can be made better would be good, though I'm not so sad about the gotos right now: it's very clearly understandable, even if you're not familiar with Go.

Reply:
I don't think it's worth getting into a discussion about the merits of goto right now, but I will say that I find it quite scary. Amongst the line noise I often miss the `goto` statement and I have no idea without grepping how many parts of the program can jump to a particular label.

Go does have `break` - could we not put this thing in a `for` and break if there's an error?

Reply:
I think you end up having to use some control flow to go back to the start after you've ran break? I'd encourage you to try it out, I wasn't able to make this nicer. You may have a much better time than I.